src/ControlSystem/Trigger.hpp

```diff
@@ -148,15 +161,53 @@ class Trigger : public DenseTrigger {
       const std::unordered_map<
           std::string,
           std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
-          measurement_timescales) {
-    return time +
-           tmpl::as_pack<ControlSystems>(
-               [&measurement_timescales, &time](auto... control_systems) {
-                 return std::min({min(
-                     measurement_timescales
-                         .at(tmpl::type_from<decltype(control_systems)>::name())
-                         ->func(time)[0])...});
-               });
+          measurement_timescales,
+      const std::unordered_map<
+          std::string,
+          std::unique_ptr<domain::FunctionsOfTime::FunctionOfTime>>&
+          functions_of_time) {
+    const double min_measure_time = tmpl::as_pack<ControlSystems>(
+        [&measurement_timescales, &time](auto... control_systems) {
+          return std::min(
+              {min(measurement_timescales
+                       .at(tmpl::type_from<decltype(control_systems)>::name())
+                       ->func(time)[0])...});
+        });
+    if (min_measure_time == std::numeric_limits<double>::infinity()) {
+      return min_measure_time;
+    }
+    const double calculated_next_measurement = time + min_measure_time;
+
+    double min_fot_expr_time = std::numeric_limits<double>::infinity();
+    for (const auto& [name, fot] : functions_of_time) {
+      (void)name;
+      min_fot_expr_time = std::min(min_fot_expr_time, fot->time_bounds()[1]);
```

Comment:
Are you sure that the expiration times are in a consistent state on all elements when they run this function?  It looks like this can run any time after the measurement timescales update, which I don't think is sequenced with respect to the main FoTs.

Reply:
Good catch. I think my new implementation catches this, but let me know if you'd like me to change it. It isn't the prettiest solution.

Reply:
Thinking about this a bit more, I think this can fail in the other direction as well:  If some element that doesn't need to supply data is more than one measurement behind the elements actually used in a measurement, it could receive the FoT update before scheduling the trigger at the update time.

Is there some other way we could obtain the update time?  The expiration time was really not intended to be used as data.

Reply:
I believe these concerns are now moot because the functions of time aren't updated at the expiration time anymore.

Reply:
Hmmm you're right. We have the `control_system::Tags::MeasurementsPerUpdate` in the cache. What if we store which measurement we are on in the trigger, and then when we are at the update time (which is `current_measurement == measurements per update`), we store the expiration time in the trigger and use that expiration time. This way there's no race condition because nothing has been updated yet because we have to actually do the measurement the functions of time can be updated. There'd have to be a special case for t=0, but that's doable. What do you think @wthrowe?

Reply:
Do we necessarily have to have done the measurement?  Someone has to have done the measurement, certainly, but a horizon-find might not wait for data from the wave zone.

What is the purpose of this roundoff check?  Can we just remove it?

Reply:
I don't think we can get rid of the roundoff check. We were having all these issues syncing up floating point numbers because the expiration time is calculated on the control system components while the next measurement time is calculated on the elements and those weren't lining up.

Given that, I just need an expiration time that's guaranteed to be valid and not subject to race conditions. This was the simplest way I could think of at the moment. If you have a better way, I'm open to suggestions, but I'd really like to get this in quickly so we can do BBH testing with it.

Reply:
But why do they have to line up?

Reply:
Originally they had to because we updated at the expiration time. But now since we update before the expiration time, I guess they don't have to line up exactly. If the measurement time is less than the expiration time by roundoff (or equal to exactly), nothing happens because everything is valid. If the measurement time is greater than the expiration time by roundoff, then it will wait for the functions of time to update which is now guaranteed to happen because we updated the previous measurement.

I've tested a BBH where I got rid of the roundoff check and it's run fine so far up to 30M which is already a lot of updates.